velox/common/memory/MemoryArbitration.h

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <vector>
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/common/base/SuccinctPrinter.h"
+
+namespace facebook::velox::memory {
+
+class MemoryPool;
+
+/// The memory arbitrator interface. There is one memory arbitrator object per
+/// memory manager which is responsible for arbitrating memory usage between the
+/// root memory pools allocated from the memory manager. Different memory
+/// arbitrator implementation enforce different arbitration policies. For
+/// example, Prestissimo needs a fixed size memory arbitration policy to enforce
```

Comment:
I wonder if this is essentially a "bug / limitation" of the current Java implementation in Presto. My understanding is that Presto over-provisions memory and enforces fixed memory cap. When the node runs out of memory, Presto picks largest memory consumer and kills it, even though it is not exceeding its cap of memory usage. I wonder if we'd be better off with having "shared" policy you describe for Sapphire + a fixed cap that cannot be exceeded even when memory is available. (BTW, let's use Presto-on-Spark and Prestissimo-on-Spark instead of Sapphire term)

Reply:
Good point. " I wonder if we'd be better off with having "shared" policy you describe for Sapphire + a fixed cap that cannot be exceeded even when memory is available. " Not sure if I fully understand what your suggestion here. Let's sync offline about this. Thanks!